src/fu-self-test.c

```diff
@@ -214,12 +214,89 @@ fu_engine_requirements_missing_func (gconstpointer user_data)
 	/* check this fails */
 	task = fu_install_task_new (NULL, component);
 	ret = fu_engine_check_requirements (engine, task,
+					    FU_ENGINE_DBUS_SENDER_INVALID,
 					    FWUPD_INSTALL_FLAG_NONE,
 					    &error);
 	g_assert_error (error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND);
 	g_assert (!ret);
 }
 
+static void
+fu_engine_requirements_client_fail_func (gconstpointer user_data)
+{
+	gboolean ret;
+	g_autoptr(XbNode) component = NULL;
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(FuEngine) engine = fu_engine_new (FU_APP_FLAGS_NONE);
+	g_autoptr(FuInstallTask) task = NULL;
+	g_autoptr(GError) error = NULL;
+	const gchar *xml =
+		"<component>"
+		"  <requires>"
+		"    <client>detach-action</client>"
+		"  </requires>"
+		"</component>";
+
+	/* make the component require one thing */
+	silo = xb_silo_new_from_xml (xml, &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (silo);
+	component = xb_silo_query_first (silo, "component", &error);
+	g_assert_no_error (error);
+	g_assert_nonnull (component);
+
+	/* check this fails */
+	task = fu_install_task_new (NULL, component);
+	ret = fu_engine_check_requirements (engine, task,
+					    FU_ENGINE_DBUS_SENDER_INVALID,
+					    FWUPD_INSTALL_FLAG_NONE,
+					    &error);
+	g_assert_error (error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED);
+	g_assert (!ret);
+}
+
+static void
+fu_engine_requirements_client_pass_func (gconstpointer user_data)
+{
+	gboolean ret;
+	g_autoptr(XbNode) component = NULL;
+	g_autoptr(XbSilo) silo = NULL;
+	g_autoptr(FuEngine) engine = fu_engine_new (FU_APP_FLAGS_NONE);
+	g_autoptr(FuInstallTask) task = NULL;
+	g_autoptr(GError) error = NULL;
+	const gchar *xml =
+		"<component>"
+		"  <requires>"
+		"    <client>detach-action</client>"
```

Comment:
do you want another case that has invalid client requirement perhaps?

Reply:
Added, good idea. I used FWUPD_ERROR_NOT_FOUND for not known, and FWUPD_ERROR_NOT_SUPPORTED for not valid, as INVALID_FILE didn't seem right.